### PR TITLE
Added all x86 SIMD codes.

### PR DESCRIPTION
**Pull request by valarauca**
[original pull request, rebased and resubmitted](https://github.com/rust-lang/rust/pull/37062)

This patch adds all x86_86 platform intrinsic. The names are compliant with those used in the GCC, ICC, and MSVC. The x86.rs file is generated via a [rust program](https://github.com/valarauca/Decode-x86-SIMD).

A ~4 months ago @sanxiyn request I add them to this commit when I got around to a submitting.

I realize this is directly modifying a generated file. But this method allows for automated updates without hand modifying writing JSON. I can extend decode-simd to include the MIPS, Aarch, Arm, POWER8, (and even NVVC+AMDGPU) platforms if that is needed/wanted.

Other commenters cc: @rkruppe @retep998

### CODE REVIEW

**Comment by valarauca** [comment]
So the current failure is one the tests. Test re-writing will be a large part of this patch, but before I start that I'd like the community to weigh on if this solution is worth adopting.

**Comment by alexcrichton** [comment]
Thanks for the PR! Would it be possible to update the JSON as well to then regenerate these files? (just to match what we've previously done here)

**Comment by valarauca** [comment]
This is where the issue arises. I don't know how to bulk generate the x86.rs from the pythonscript + json files. As far as I can tell you get 1 x86.rs PER json file, then hand merge them(?). And if we continue the pattern of 1 json file per feature set, this patch will require around a dozen new json files.

Really my main goal was to completely avoid the json files, or even replace them. They're incredibly frustrating to work with in practice.

**Comment by sanxiyn** [comment]
I think we should give up on JSON, and that is what this PR is about.

**Comment by valarauca** [comment]
@sanxiyn should I move forward and start extending decode SIMD to Aarch64, Arm, and MIPS? 

As this isn't a _total_ replacement for the Json/Python system I understand the hesitation with accepting it.

@alexcrichton For AArch+Arm should I attempt to persevere the current names/symbols? Or should I not be concerned if they're changed?

**Comment by alexcrichton** [comment]
@valarauca ah ok, interesting! Sorry but I was unaware about that being the point of this PR.

Replacing the JSON files sounds good to me so long as we do it in a methodic fashion. Would it be possible to jettison them entirely and replace them with [your script](https://github.com/valarauca/Decode-x86-SIMD) you've got? It'd be great to prove this out with the other platforms as well!

**Comment by valarauca** [comment]
Okay so for writing tests. What is the difference between `compile-fail`, `run-make`, and `run-pass`? Because I'll need to edit these files.

For a primary test to _prove_ our SIMD functionality should the _script_ also have a feature to create a separate crate with a `build.rs` directive for linking against the GCC to _prove_ out it's functionality?

**Comment by valarauca** [comment]
I removed the `deny(warnings)` take. It caused a TON of headaches in testing as ARM/x86 don't share some types. And ARM would throw an error, but x86 wouldn't. Then for some types x86 would throw errors where arm wouldn't.

**Comment by alexcrichton** [comment]
The compile-fail tests are ones that are expected to have a compile failure and have an assertion to what that error is. The run-make tests are just tests that need a `Makefile` to run (e.g. a kitchen sink for "this can't run in any other test suite" tests). Ideally run-make tests are avoided if possible. The run-pass test suite asserts that the tests compile and execute successfully.

It's ok to not add exhaustive testing to this though, we're likely not quite prepared for that.

**Comment by valarauca** [comment]
Okay. I just had to clean up a few of the tests because they were using names/symbols that no longer exist. The `u16` datatype isn't technically supported on intel intristics so I removed it, but the compile tests were using it.

**Comment by valarauca** [comment]
Hey @alexcrichton I need your help with this one

`{"message":"intrinsic argument 1 has wrong type: found vector with length 8, expected length 16","code":{"code":"E0442","explanation":"\nIntrinsic argument(s) and/or return value have the wrong type.\nErroneous code example:\n\n```compile_fail,E0442\n#![feature(repr_simd)]\n#![feature(platform_intrinsics)]\n\n#[repr(simd)]\nstruct i8x16(i8, i8, i8, i8, i8, i8, i8, i8,\n             i8, i8, i8, i8, i8, i8, i8, i8);\n#[repr(simd)]\nstruct i32x4(i32, i32, i32, i32);\n#[repr(simd)]\nstruct i64x2(i64, i64);\n\nextern \"platform-intrinsic\" {\n    fn x86_mm_adds_epi16(x: i8x16, y: i32x4) -> i64x2;\n    // error: intrinsic arguments/return value have wrong type\n}\n```\n\nTo fix this error, please refer to the function declaration to give\nit the awaited types. Example:\n\n```\n#![feature(repr_simd)]\n#![feature(platform_intrinsics)]\n\n#[repr(simd)]\nstruct i16x8(i16, i16, i16, i16, i16, i16, i16, i16);\n\nextern \"platform-intrinsic\" {\n    fn x86_mm_adds_epi16(x: i16x8, y: i16x8) -> i16x8; // ok!\n}\n```\n"},"level":"error","spans":[{"file_name":"/build/src/test/run-pass/simd-upgraded.rs","byte_start":866,"byte_end":915,"line_start":24,"line_end":24,"column_start":13,"column_end":62,"is_primary":true,"text":[{"text":"            fn x86_mm_min_epi16(x: i16x8, y: i16x8) -> i16x8;","highlight_start":13,"highlight_end":62}],"label":null,"suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"intrinsic argument 2 has wrong type: found vector with length 8, expected length 16","code":{"code":"E0442","explanation":"\nIntrinsic argument(s) and/or return value have the wrong type.\nErroneous code example:\n\n```compile_fail,E0442\n#![feature(repr_simd)]\n#![feature(platform_intrinsics)]\n\n#[repr(simd)]\nstruct i8x16(i8, i8, i8, i8, i8, i8, i8, i8,\n             i8, i8, i8, i8, i8, i8, i8, i8);\n#[repr(simd)]\nstruct i32x4(i32, i32, i32, i32);\n#[repr(simd)]\nstruct i64x2(i64, i64);\n\nextern \"platform-intrinsic\" {\n    fn x86_mm_adds_epi16(x: i8x16, y: i32x4) -> i64x2;\n    // error: intrinsic arguments/return value have wrong type\n}\n```\n\nTo fix this error, please refer to the function declaration to give\nit the awaited types. Example:\n\n```\n#![feature(repr_simd)]\n#![feature(platform_intrinsics)]\n\n#[repr(simd)]\nstruct i16x8(i16, i16, i16, i16, i16, i16, i16, i16);\n\nextern \"platform-intrinsic\" {\n    fn x86_mm_adds_epi16(x: i16x8, y: i16x8) -> i16x8; // ok!\n}\n```\n"},"level":"error","spans":[{"file_name":"/build/src/test/run-pass/simd-upgraded.rs","byte_start":866,"byte_end":915,"line_start":24,"line_end":24,"column_start":13,"column_end":62,"is_primary":true,"text":[{"text":"            fn x86_mm_min_epi16(x: i16x8, y: i16x8) -> i16x8;","highlight_start":13,"highlight_end":62}],"label":null,"suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"intrinsic return value has wrong type: found vector with length 8, expected length 16","code":{"code":"E0442","explanation":"\nIntrinsic argument(s) and/or return value have the wrong type.\nErroneous code example:\n\n```compile_fail,E0442\n#![feature(repr_simd)]\n#![feature(platform_intrinsics)]\n\n#[repr(simd)]\nstruct i8x16(i8, i8, i8, i8, i8, i8, i8, i8,\n             i8, i8, i8, i8, i8, i8, i8, i8);\n#[repr(simd)]\nstruct i32x4(i32, i32, i32, i32);\n#[repr(simd)]\nstruct i64x2(i64, i64);\n\nextern \"platform-intrinsic\" {\n    fn x86_mm_adds_epi16(x: i8x16, y: i32x4) -> i64x2;\n    // error: intrinsic arguments/return value have wrong type\n}\n```\n\nTo fix this error, please refer to the function declaration to give\nit the awaited types. Example:\n\n```\n#![feature(repr_simd)]\n#![feature(platform_intrinsics)]\n\n#[repr(simd)]\nstruct i16x8(i16, i16, i16, i16, i16, i16, i16, i16);\n\nextern \"platform-intrinsic\" {\n    fn x86_mm_adds_epi16(x: i16x8, y: i16x8) -> i16x8; // ok!\n}\n```\n"},"level":"error","spans":[{"file_name":"/build/src/test/run-pass/simd-upgraded.rs","byte_start":866,"byte_end":915,"line_start":24,"line_end":24,"column_start":13,"column_end":62,"is_primary":true,"text":[{"text":"            fn x86_mm_min_epi16(x: i16x8, y: i16x8) -> i16x8;","highlight_start":13,"highlight_end":62}],"label":null,"suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}
{"message":"aborting due to 3 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":null}`

This occured in `/rust/src/test/run-pass/simd-upgraded.rs` I have an `i16x8` type defined in `/rust/src/librustc_platform_intrinsics/lib.rs`, but in the error message there isn't anything mentioning `i16x8` as a possible size for a 16byte tuple type.

File: src/librustc_platform_intrinsics/lib.rs [hunk 1]
```diff
@@ -37,60 +36,41 @@ pub enum IntrinsicDef {
...
+static VOID: Type = Type::Void;
+
 static I8: Type = Type::Integer(true, 8, 8);
 static I16: Type = Type::Integer(true, 16, 16);
 static I32: Type = Type::Integer(true, 32, 32);
 static I64: Type = Type::Integer(true, 64, 64);
-static U8: Type = Type::Integer(false, 8, 8);
-static U16: Type = Type::Integer(false, 16, 16);
-static U32: Type = Type::Integer(false, 32, 32);
-static U64: Type = Type::Integer(false, 64, 64);
```
Comment (alexcrichton) [inline src/librustc_platform_intrinsics/lib.rs]: Does this mean that all unsigned simd intrinsics were removed? Perhaps that's the cause of the error message?

**Comment by alexcrichton** [comment]
Thinking a bit more about this, there's also a few points about the existing system I forgot initially:

* The python script currently has a `--help` flag which should help generate json definitions if they need to be imported, so in theory the syntax isn't *too* alien.
* The json files are also used to generate the `extern` blocks in *other* repositories like https://github.com/huonw/simd, which allows for not only easy compiler imports but also consuming this downstream as well.

I wonder though if we checked in a Rust script if it could help both importing to the compiler and exporting to external libraries as well? (e.g. dual mode-esque)

**Comment by valarauca** [comment]
Re-writing/patching Huonw's SIMD library was my next action item after this stabilized. Really that library should be part of the compiler at it is describing core types. You don't important a `boost::` library for `_m128` in C++, yet this is the current model rust/cargo uses. 

But I do recognize it is easier to have a feature start out as a crate, then move that crate into the compiler. Rather then have an extremely buggy section of the compiler.

Furthermore some SIMD operations aren't easily quantifiable as integer/vector operations. AArch/Arm/Intel CRC, Intel SHA1/SHA2/AES, and the various SSE4.2 String operations. 

Lastly I read the `-help` text. The differences between `S` and `s` is frustrating for writing by hand. The Pointer syntax is also rather annoying.

File: src/librustc_platform_intrinsics/lib.rs (hunk 1 above)
Comment (valarauca) [inline src/librustc_platform_intrinsics/lib.rs]: The error is happening with the type `i16x8`. 

As I said previously

This occured in `/rust/src/test/run-pass/simd-upgraded.rs` I have an `i16x8` type defined in `/rust/src/librustc_platform_intrinsics/lib.rs`, but in the error message there isn't anything mentioning i16x8 as a possible size for a 16byte tuple type.

Comment (alexcrichton) [inline src/librustc_platform_intrinsics/lib.rs]: Why were these types removed? This is removing a number of unsigned simd intrinsics from tests below. I was just hazarding a guess that it was related to the test failure, I'm not really sure what's going on there.

**Comment by alexcrichton** [comment]
Note that the story around SIMD in Rust is very much still a work in progress. We've got @huonw's `simd` library as an excellent starting point and is where we'll likely expand from moving forward. The libs team is set to discuss SIMD in general in the near future so we should have more info on that soon as well.

In that sense I'd hold off large-scale rewrites for the time being, but fleshing out all the intrinsics in the compiler sounds great to me, so I'd still be interested in moving forward with this. To me the remaining issues here are:

* We should check in the generation script to this repo so it can be rerun if necessary.
* Tests need to become green

Other than that though I'm comfortable landing!

**Comment by valarauca** [comment]
Sounds good. Where will that discussion take place? Because I'd like to chime in on it. 

Either way I hope to have the tests green this weekend. Emergency work travel slowed my commits heavily :|

File: src/librustc_platform_intrinsics/lib.rs (hunk 1 above)
Comment (valarauca) [inline src/librustc_platform_intrinsics/lib.rs]: From the error message it seems to be stating there is no 16byte width intrinsic defined with a type signature of `i16`. While there is `i8x16`, `i32x4`, and `i64x2`, these fail the type check. I'm digging into it today.

**Comment by alexcrichton** [comment]
The libs team will discuss SIMD soon (hopefully!) and we'll post all our thoughts to an [internals thread](https://internals.rust-lang.org/) afterwards.

**Comment by alexcrichton** [comment]
That being said I think it's fine to land this in the meantime, given:

* It's just expanding the number of exposed intrinsics
* They're all still unstable
* We'll still have a script to generate everything

File: src/librustc_platform_intrinsics/lib.rs (hunk 1 above)
Comment (rkruppe) [inline src/librustc_platform_intrinsics/lib.rs]: > The LLVM intrinsic definitions themselves don't use unsigned values, all their type signatures are signed except for floating point values. Therefore the generation program is only generating signed intrinsics. 

It's not that the LLVM intrinsics have signed or unsigned integers in their signatures, LLVM integer types simply don't have a signedness, instead they put it in the operations. But that doesn't mean the Rust interface to those intrinsics should ignore Rust's richer type systems. Requiring wrapper crates to insert lots of ugly `transmute` would be a pointless complication IMHO, and would potentially generate worse code (in debug mode at least). To me this points to a weakness of the "generate everything from LLVM definitions" approach more than anything else.

Comment (valarauca) [inline src/librustc_platform_intrinsics/lib.rs]: >That doesn't mean the Rust interface to those intrinsics should ignore Rust's richer type systems. Requiring wrapper crates to insert lots of ugly `transmute` would be a pointless complication IMHO,

Nothing is ignoring Rust's type system. Rust ignores Rust's type system at the `extern platform-instrict` layer. This is where you can inject `u64x2( u64,u64);` unsigned types perfectly safely then any interaction will be type safe to the `u64x2` type. 

If is is that big of a deal I'll re-add the unsigned types.

**Comment by valarauca** [comment]
Closing will submit JSON changes this weekend.

**Pull request closed without merging**